### PR TITLE
[2/2] Remove training loop force calling early stopping callback

### LINKED ISSUE #7033: Support early stopping during training inside of the early stopping callback
## 🚀 Feature

The ability to early stop during training, as controlled by the callback instead of the training loop.

### Motivation

This simplifies the training loop from manually running these callback hooks if the validation epoch won't be run: https://github.com/PyTorchLightning/pytorch-lightning/blob/5bd3cd5f712b65d38812b27cf957261bb06b83c5/pytorch_lightning/trainer/training_loop.py#L152-L159

Similar to flexibility with checkpointing callbacks, this can eventually enable users to specify separate early stop criteria for both training and validation.

### Pitch

Add an `check_on_train_epoch_end` flag to the callback constructor. See https://github.com/PyTorchLightning/pytorch-lightning/pull/6944/ for a sketch

This flag controls whether we check during training or validation. Because the monitor metric may be in training but not validation, or vice versa, this flag makes the check across these two hooks mutually exclusive.

```
    def on_train_epoch_end(self, trainer, pl_module, outputs) -> None:
        if not self._on_train_epoch_end or self._should_skip_check(trainer):
            return
        self._run_early_stopping_check(trainer)

    def on_validation_end(self, trainer, pl_module):
        if self._on_train_epoch_end or self._should_skip_check(trainer):
            return
        self._run_early_stopping_check(trainer)
```

For parity with existing behavior, by default this flag will be False by default. With this feature enabled, users can specify their callback for early stopping during training as such:

```
stop = EarlyStopping(monitor='abc', min_delta=0.1, patience=0, check_on_train_epoch_end=True)
```

Users could then create multiple such callbacks:
```
train_stop = EarlyStopping(monitor='abc', min_delta=0.1, patience=0, check_on_train_epoch_end=True)
val_stop = EarlyStopping(monitor='val_loss', min_delta=0.5, patience=3)
trainer = Trainer(...., callbacks=[train_stop, val_stop], ...)
trainer.fit(...)
```

### Alternatives

Keep as is

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7033 
Part 2 - this depends on #6944

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -129,20 +133,28 @@ def on_save_checkpoint(self, trainer, pl_module, checkpoint: Dict[str, Any]) ->
...
-        if trainer.state != TrainerState.FITTING or trainer.sanity_checking:
+        return trainer.state != TrainerState.FITTING or trainer.sanity_checking
+
+    def on_train_epoch_end(self, trainer, pl_module, outputs) -> None:
+        if not self._check_on_train_epoch_end or self._should_skip_check(trainer):
+            return
+        self._run_early_stopping_check(trainer)
+
+    def on_validation_end(self, trainer, pl_module) -> None:
+        if self._check_on_train_epoch_end or self._should_skip_check(trainer):
```
Comment (tchaton) [inline pytorch_lightning/callbacks/early_stopping.py:152]: Should we run it for both validation and training ?

Comment (tchaton) [inline pytorch_lightning/callbacks/early_stopping.py:152]:
```suggestion
        if self._should_skip_check(trainer):
```

Comment (ananthsub) [reply pytorch_lightning/callbacks/early_stopping.py:152]: No, as the monitor metric might not be available in both training and validation. This lets people mix and match better, similar to what we are doing with checkpointing.

Comment (awaelchli) [reply pytorch_lightning/callbacks/early_stopping.py:152]: side fact: We actually had at one point model checkpoint running on both training and val epoch end, which lead to writing the checkpoint twice, once with the old value for the val metric and once with the new one. And then it was wrong of course when val_check_interval != 1.

**Review by Borda — COMMENTED** [review]
just to clarify, this is a cumulative change including #6944 correct?

**Comment by ananthsub** [comment]
> just to clarify, this is a cumulative change including #6944 correct?

yes thats correct

**Comment by ananthsub** [comment]
@carmocca @awaelchli @tchaton @Borda mind taking a look?

**Review by Borda — APPROVED** [review]
lgtm

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -548,7 +548,7 @@ def training_step(self, batch, batch_idx):
             return output
 
     model = TestModel()
-    early_stop = EarlyStopping(monitor="loss", patience=0)
+    early_stop = EarlyStopping(monitor="loss", patience=0, check_on_train_epoch_end=True)
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:551]: Just want to bring awareness, this changes the behavior for users. 
A user that has monitor on a training metric now needs to set this argument after upgrading, right? Let's make this clear in the changelog, in the "Changed" section?

And are we good to include this in 1.3 during feature freeze here? I assume so since milestone is set to 1.3 but just to make it clear to everyone.

**Review by awaelchli — APPROVED** [review]
Nice to finally see this TODO getting resolved!

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (ananthsub) [reply tests/trainer/test_trainer.py:551]: Added to the changelog. @edenlightning @tchaton wdyt for 1.3?

Comment (ananthsub) [reply tests/trainer/test_trainer.py:551]: @awaelchli fwiw this is pretty new behavior: https://github.com/PyTorchLightning/pytorch-lightning/pull/5208

but definitely, it's a change compared to before and i dont think there's a way to make this backward compatible. logging a warning also can get lost. keeping this around longer blocks the general loop refactor we want to do

Comment (awaelchli) [reply tests/trainer/test_trainer.py:551]: yes, it can't be made compatible. I just want to push for a better changelog. 
Thanks @ananthsub
